Migrate to Capacitor 3

Nevvulo opened the pull request:
This PR updates Capacitor from version 2 to version 3, and updates the plugin implementation and configuration to work with Capacitor 3.

* changed iOS deployment target from iOS 11 to iOS 12
* removes unused dependencies in Podspec file
* change version from 0.4.1 to 0.5.0
* update Android annotations in plugin implementation
* changed `call.error` -> `call.reject` and `call.success` -> `call.resolve`
* removed any instances of `call.hasOption` (`hasOption` has been deprecated)
* update web implementation for new registration process

Nevvulo committed:
Change iOS deployment target to 12.0, remove unused dependencies

The podspec file has been modified to match the one at https://github.com/capacitor-community/firebase-analytics

Nevvulo committed:
Add required Android implementations to gradle build file

Nevvulo committed:
Upgrade Capacitor from v2 to v3

Nevvulo committed:
Update initialization of web plugin to Capacitor v3

registerWebPlugin has been deprecated in v3, the preferred way to initialize a web plugin is registerPlugin with a custom web implementation from the plugin class

Nevvulo committed:
Update tsconfig target to recommended es2017

Nevvulo committed:
Update iOS plugin implementation to Capacitor 3 best practices

call.success and call.error have been deprecated, call.hasOption has also been deprecated

Nevvulo committed:
Update Pods + Podfile for plugin

Nevvulo committed:
Implement new Capacitor v3 annotations for Android impl.

@NativePlugin was deprecated, the preferred new annotation is @CapacitorPlugin

Nevvulo committed:
Fix warning in web console about config

The config for the plugin is set up when using registerPlugin and doesn't need to be supplied in the constructor

Nevvulo committed:
Remove unnecessary force unwrap, add default value on instance ID

gugahoi commented on android/src/main/java/com/getcapacitor/community/firebaseanalytics/FirebaseAnalytics.java:
```diff
@@ -3,21 +3,30 @@
+      alias = "access_network_state"
+    ),
+    @Permission(strings = { Manifest.permission.INTERNET }, alias = "internet"),
+    @Permission(
+      strings = { Manifest.permission.WAKE_LOCK },
+      alias = "wake_lock"
+    ),
```
Curious on if this is still necessary and the impact this has. Do you know?

Nevvulo replied:
I'm not entirely sure if this is necessary or what the impact is: [the original repo currently has all 3 permissions still defined](https://github.com/capacitor-community/firebase-analytics/blob/master/android/src/main/java/com/getcapacitor/community/firebaseanalytics/FirebaseAnalytics.java#L25-L28) so perhaps leaving this is best for now, but I'll also investigate if we can safely remove this without impact

gugahoi replied:
I agree, prob best not to change for now but would be good to know if we can remove it in future.

gugahoi commented on ios/Plugin/Plugin.swift:
```diff
@@ -20,51 +20,47 @@ public class FirebaseAnalytics: CAPPlugin {
-        call.success([
-            "instanceId": instanceId
+        call.resolve([
+            "instanceId": instanceId ?? "0"
```
Not so sure about the `0` here, is this in line with previous behaviour?

Nevvulo replied:
This is not in line with the previous behaviour, do you think rejecting the call if `instanceId` is `nil` is appropriate? I've only made this change because there's a warning here as we need to unwrap `instanceId` from `Analytics.appInstanceID()` or provide a default value

Nevvulo replied:
The Android implementation currently throws an error if the app instance ID is unable to be retrieved so I'll match that behaviour for now unless we want something else

gugahoi commented on src/definitions.ts:
```diff
@@ -1,10 +1,10 @@
-    FirebaseAnalytics: FirebaseAnalyticsPlugin;
-  }
-}
+// declare module "@capacitor/core" {
+//   interface PluginRegistry {
+//     FirebaseAnalytics: FirebaseAnalyticsPlugin;
+//   }
+// }
```
Can we delete this?

gugahoi reviewed (changes requested):
Do we also need to update the Android implementation to use `resolve`/`reject`?

Nevvulo committed:
Update success/error calls to resolve/reject on Android impl.

Nevvulo committed:
Remove unneeded comment from web definitions

Nevvulo committed:
Reject call on iOS if app instance ID is unavailable to match Android impl.

Nevvulo commented:
I've updated the Android implementation to use `resolve/reject` calls, updated iOS implementation to match Android when app instance ID is missing, and removed the unnecessary commented code in web type definitions

gugahoi reviewed (commented):
🙏
